components/resource/objectcache.hpp

```diff
@@ -32,71 +36,161 @@ namespace osg
 
 namespace Resource {
 
-class ObjectCache : public osg::Referenced
+template <typename KeyType>
+class GenericObjectCache : public osg::Referenced
 {
     public:
 
-        ObjectCache();
+        GenericObjectCache()
+            : osg::Referenced(true) {}
 
         /** For each object in the cache which has an reference count greater than 1
           * (and therefore referenced by elsewhere in the application) set the time stamp
           * for that object in the cache to specified time.
           * This would typically be called once per frame by applications which are doing database paging,
           * and need to prune objects that are no longer required.
           * The time used should be taken from the FrameStamp::getReferenceTime().*/
-        void updateTimeStampOfObjectsInCacheWithExternalReferences(double referenceTime);
+        void updateTimeStampOfObjectsInCacheWithExternalReferences(double referenceTime)
+        {
+            OpenThreads::ScopedLock<OpenThreads::Mutex> lock(_objectCacheMutex);
+            for(typename ObjectCacheMap::iterator itr=_objectCache.begin(); itr!=_objectCache.end(); ++itr)
+            {
+                // if the timestamp is yet to be initialized, update it as well
```

Comment:
If the movement of declaration to the definition is unavoidable, you should still copy-paste the whole comment as it is from the implementation file, bzzt has omitted the original comments for whatever reason so I adjusted and included all of them.